### PR TITLE
timelock: handle log processing shutdown and Listen shutdown

### PR DESCRIPTION
**Pull request by MStreet3**
This PR shuts down the log processing goroutine if both log channels are closed.  If both log channels close, then the context is canceled which forces a shutdown.

This PR also cancels the scheduling goroutine in the event that the processing goroutine shuts down.  Once the Listen method starts to shutdown the worker, it now waits for all goroutines to finish with only a five second grace period.

### CODE REVIEW

File: pkg/timelock/timelock.go [hunk 1]
```diff
@@ -154,16 +155,18 @@ func (tw *Worker) Listen() error {
 	select {
 	case <-ctxwc.Done():
 	case <-processingDone:
+		cancel()
```
Comment (MStreet3) [inline pkg/timelock/timelock.go:158]: call cancel here to stop the scheduling routine

File: pkg/timelock/timelock.go [hunk 2]
```diff
@@ -154,16 +155,18 @@ func (tw *Worker) Listen() error {
...
 
 	tw.logger.Info().Msg("shutting down timelock-worker")
 	tw.logger.Info().Msg("dumping operation store")
 	tw.dumpOperationStore(time.Now)
 
 	// Wait for all goroutines to finish.
-	<-historyDone
-	<-newDone
-	<-schedulingDone
+	shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```
Comment (MStreet3) [inline pkg/timelock/timelock.go:166]: create a new context as the previous one is either done or was canceled

File: pkg/timelock/timelock.go [hunk 3]
```diff
@@ -273,26 +276,47 @@ func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (<-chan struct{},
 // processLogs is implemented as a fan-in for all the logs channels, merging all the data and handling logs sequentially.
 // This function is thread safe.
 func (tw *Worker) processLogs(ctx context.Context, oldLog, newLog <-chan types.Log) <-chan struct{} {
-	done := make(chan struct{})
+	var (
+		done, newDone, oldDone = make(chan struct{}), make(chan struct{}), make(chan struct{})
+		ctxwc, cancel          = context.WithCancel(ctx)
```
Comment (MStreet3) [inline pkg/timelock/timelock.go:281]: create a child context we can cancel when both log channels are closed

File: pkg/timelock/timelock.go [hunk 4]
```diff
@@ -273,26 +276,47 @@ func (tw *Worker) retrieveHistoricalLogs(ctx context.Context) (<-chan struct{},
...
 	go func() {
 		defer close(done)
 
 		for {
 			select {
-			case log := <-newLog:
-				if err := tw.handleLog(ctx, log); err != nil {
+			case log, open := <-newLog:
+				if !open {
+					close(newDone)
```
Comment (MStreet3) [inline pkg/timelock/timelock.go:298]: if newLog is closed, signal new logs are done and set the channel to nil so this branch never is selected again

**Review by fbac — APPROVED** [review]

**Pull request merged**